Fluffyemily/fastlane integration

fluffyemily opened the pull request:
Fastlane for aurora builds

Setup fastlane and all the associated tools by running:

`./fastlane/scripts/setup.sh`

once everything is installed create your aurora build using 

`fastlane <aurora | l10n> [branch:<branch to build from>] [base_branch:<branchToBaseNewBranchFrom>] [localise:<true | false>] build:<buildNumber> version:<versionNumber> username:<apple_id>`

your build will be output as `builds/firefox-ios-<versionNumber>-<buildNumber>-<timestamp>.ipa`

and everything will have been checked in to a new branch (but not pushed)

Resolves bugs

https://bugzilla.mozilla.org/show_bug.cgi?id=1217855
https://bugzilla.mozilla.org/show_bug.cgi?id=1212327
https://bugzilla.mozilla.org/show_bug.cgi?id=1212037
https://bugzilla.mozilla.org/show_bug.cgi?id=1217858

rnewman commented:
A promising start:

```
$ ./fastlane/scripts/setup.sh
./fastlane/scripts/setup.sh: line 19: print: command not found
﻿﻿Homebrew already installed, updating
Updated Homebrew from 8ef29e6c to 56ef538f.
```

rnewman commented:
I'm using `zsh`, FWIW.

rnewman commented:
```
Error: imagemagick-6.9.0-10 already installed
To install this version, first `brew unlink imagemagick'
Carthage already installed
Installing fastlane.
Error: No available formula with the name "fastlane"
==> Searching for similarly named formulae...
Error: No similarly named formulae found.
==> Searching taps...
Error: No formulae found in taps.
./fastlane/scripts/setup.sh: line 84: fastlane: command not found
```

sleroux commented on ClientUITests/snapshot/MarketingSnapshotTests.swift:
```diff
@@ -0,0 +1,118 @@
+//
```
MPL 2.0 License

fluffyemily commented:
@rnewman Updated the setup script.

sleroux commented:
I just ran the fastlane build script after your fixes and got stuck on sigh doing the code signing since it's connected to your email. Is there a way to configure this to another email?

fluffyemily commented:
@sleroux for now, open up fastlane/Appfile and edit the apple_id to be your Apple ID. We're working on getting a shared AppleID for build purposes.

sleroux commented:
So close! I got it to build using my Apple ID but got a bunch of errors such as:

❌  error: failed to remove /Users/sleroux/Library/Developer/Xcode/DerivedData/Client-cbhfzrpvkymhhqgrbaozbigqzuph/Build/Intermediates/ArchiveIntermediates/FennecAurora/IntermediateBuildFilesPath/UninstalledProducts/iphoneos/ViewLater.appex/id.lproj/InfoPlist.strings: “InfoPlist.strings” couldn’t be removed.

fluffyemily commented:
sleroux: Can you check what the Info.plist value is in the ViewLater target Build Settings?

fluffyemily commented:
@sleroux actually, can you email me the whole error trace?

fluffyemily commented:
OK, this seems to be an error caused when you run the build against a branch that has already been localized previously. Will take a look.

fluffyemily commented:
New version.

Please delete all previously created branches before executing this time.

also, please include extra command line arg: username:<apple_id>, so execute with

`fastlane aurora branch:fluffyemily/fastlane-integration build:<buildNumber> version:<versionNumber> username:<apple_id>`

sleroux commented:
Got a different error codesigning Alamofire :(
> The following build commands failed:
CodeSign /Users/sleroux/Library/Developer/Xcode/DerivedData/Client-cbhfzrpvkymhhqgrbaozbigqzuph/Build/Intermediates/ArchiveIntermediates/FennecAurora/IntermediateBuildFilesPath/UninstalledProducts/iphoneos/Alamofire.framework

I removed the previous two branches that were created by fastlane before running the script

fluffyemily commented:
@sleroux When it fails, it usually outputs an error log. Could you send it to me please?

sleroux commented:
@fluffyemily Sent over the logs

fluffyemily commented:
@sleroux Thanks for the logs. The reason this is failing is because of our carthage issues. If you could open the schemes and delete the Alamofire OSX, tvOS & watchOS schemes that should solve it. I thought I had submitted a patch to do that, but it doesn't seem to have propagated properly.

If we were using carthage correctly, it would only be installing the iOS library (which also causes other issues elsewhere). If you still get issues, you have to rename the product for alamofire OSX, tvOS and watchOS to something other than Alamofire. Sadly all the targets get built with the same name, so when we link it's pot luck whether we link to the iOS one or one of the others. Again, using carthage correctly would resolve these issues.

st3fan commented:
I'm having trouble at the very beginning of the build https://gist.github.com/st3fan/ba3f818b2dd6a341d82d

st3fan commented:
Sorry if this is nitpicking but I would prefer if the branch name and directory name did not contain funny characters like `(` and `)` - how about changing:

> `firefox-ios-ios-aurora-1.3(1301)`

to:

> `firefox-ios-ios-aurora-1.3-1301`

There is also a double `ios` in there.

st3fan commented:
The import locales step for an l10n build should not use the `--skip-incomplete` option. That is only for app store builds. For all other builds we want to see all locales, including the ones in progress / not complete.

fluffyemily commented:
@st3fan @sleroux I've updated to use a forked version of Alamofire that doesn't have the OSX & watchOS schemas and targets and rebased against master. Please ensure you run ./checkout.sh on the fluffyemily/fastlane-integration branch before running again.

I've also updated the `branch:` arg to `base_branch` and changed the name of the build branch to remove the invalid chars.

From now on call like this:
`fastlane aurora base_branch:fluffyemily/fastlane-integration build:<buildNumber> version:<versionNumber> username:<apple_id>`

sleroux commented:
I ran this after running ./checkout.sh again and I'm getting the missing localise.sh script @st3fan mentioned above but I didn't get the ruby error:

https://gist.github.com/sleroux/ba8e51b54147b12bcdf6

fluffyemily commented:
@sleroux is the `scripts/localise.sh` directory present in your repo? It's here in the branch, so it's odd that it can't find it.

sleroux commented:
Looks like it's there but it's looking in the ../scripts/localise.sh path. Maybe it's relative to the current directory so it should be ./scripts/localisesh?

fluffyemily commented:
@sleroux if it is, that's new. It has always executed as if it was inside the `fastlane` dir not the `root` dir in the past.

I've just created a while new clone and tested this, and it found the localise.sh script. Are you sure your fastlane-integration branch is properly up to date?

fluffyemily commented:
Ah, did you have an existing branch by any chance? If so, delete it first. It's possible that when it checks out that existing branch it will lose the fastlane changes.

fluffyemily commented:
I've made a couple of new changes.

Each time you build, if you don't provide a `branch:` argument it will create a new branch, rather than using an existing branch for the same build/version combo. This will resolve the issues regarding new fastlane scripts failing when moving to an existing branch with an older version of fastlane

Also, if you don't provide a `base_branch:` argument, instead of defaulting to building a branch off master, it will build a branch off the current executing branch

I have also added a `localise: true` argument. If this is present, the branch will be localised, otherwise it is not. This is so you can do a l10n or aurora build with `localise:true` to create a localised branch, then use the same localised branch with no `localise` arg to create a beta or release build (i.e. another build that is localised _exactly_ the same way as a previous build).

I have just discovered another fastlane action in their Advanced documentation: `import_from_git`. This allows you to specify a default fastlane implementation and then import that into your project during builds. I'm going to investigate whether this would enable us to have our fastlane implementation in another repository and therefore all branches would get the latest build tools regardless of when they were cut. Thoughts?

sleroux commented:
> I'm going to investigate whether this would enable us to have our fastlane implementation in another repository and therefore all branches would get the latest build tools regardless of when they were cut. Thoughts? 
This sounds awesome! Doesn't really make sense to version our build chain alongside the app. Would we need to make changes to both repos in lockstep at some point like updating dependencies?

sleroux commented:
I checked out the latest and removed any previous branches that we made and getting a different error this time:

https://gist.github.com/sleroux/3293e1973ecfba9cddc9

The command I ran was:

fastlane aurora branch:fluffyemily/fastlane-integration build:1 version:1.0.0 username:sleroux@mozilla.com

fluffyemily commented:
try

`fastlane aurora localise:true build:1 version:1.0.0 username:sleroux@mozilla.com`

while on the fluffyemily/fastlane-integration branch

fluffyemily commented:
also, check your project. Have you run `./checkout.sh` and fetched the forked copy of Alamofire?

sleroux commented:
Yay it worked! I ended up having to git clean -fdx (any cache files locally), cleaning the build folder from Xcode, and re-checking everything out :+1:

fluffyemily commented:
@sleroux yay! Yeah, there is a command that causes the build to fail if the git repo isn't clean which I shall add as an extra precaution.

Glad it finally worked for someone else!

fluffyemily commented:
@sleroux @st3fan I've updated this to move all the fastlane stuff to a new repo, found here: https://github.com/mozilla/firefox-ios-build-tools. 

If you give me a +1 on this branch, I'll rebase the hell out of it (make it 1 commit) and merge it in and we can continue developing new build tools in the new repo.

I'm guessing we'd want this on 1.x as well as master?

sleroux commented:
Would it be possible to split out the marketing snapshot work into it's own PR/branch and keep this one just for fastlane building? You can probably smoosh all of the commits now, reset the commit, then add all the parts except for the marketing snapshots stuff so we'll have a single commit for just the fastlane build integration. After that I can give it a look before merging it in.

As far as getting it in 1.x, let's merge it into master for now and set the target milestone flag to 1.3. I'll probably do an uplift pass tomorrow for another TF build so I'll uplift it if it's in master along with the other stuff that needs to go in in one go.
